### PR TITLE
Implement simple wrapper over multiple TestProcessorNodes

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR adds the `TestNetwork` utility struct.

Can be used to replace this:
```
	advertiser := integrationTests.CreateMessengerWithKadDht("")
	_ = advertiser.Bootstrap(0)

	numOfShards := 1
	nodesPerShard := 1
	numMetaNodes := 10

	nodes := integrationTests.CreateNodes(
		numOfShards,
		nodesPerShard,
		numMetaNodes,
		integrationTests.GetConnectableAddress(advertiser),
	)
	integrationTests.DisplayAndStartNodes(nodes)

	defer func() {
		_ = advertiser.Close()
		for _, n := range nodes {
			_ = n.Messenger.Close()
		}
	}()
```
with this:
```
	net := integrationTests.NewTestNetwork(t)
	net.NumShards = 1
	net.NodesPerShard = 1
	net.NodesInMetashard = 10

	net.Start()
	defer func() {
		net.Close()
	}()
```

Nodes are available at `net.Nodes`.

Minting all the nodes: 
```
net.MintNodeAccounts(bigInt)
// or
net.MintNodeAccountsInt64(intValue)
```

Stepping over rounds:
```
net.Step()
net.Steps(multipleRounds)
```
`Step()` and `Steps()` do not return the round and nonce, because they are available at `net.Round` and `net.Nonce`, automatically updated at each step.

Please add more methods and make `TestNetwork` as useful as possible, to avoid copy-pasting initialization code for tests.

Any test-specific utility functions may be written to accept a `TestNetwork` instead of parameters such as `nodes`, `idxProposers`, `nonce`, `round` etc.

### CODE REVIEW

File: integrationTests/testNetwork.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	Proposers         []int
+	Advertiser        p2p.Messenger
+	AdvertiserAddress string
+	Round             uint64
+	Nonce             uint64
+	T                 *testing.T
+}
+
+func NewTestNetwork(t *testing.T) *testNetwork {
+	// TODO replace testing.T with testing.TB everywhere in the integrationTest utilities
```
Comment (iulianpascalau) [inline integrationTests/testNetwork.go:25]: :+1:

File: integrationTests/testNetwork.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	return &testNetwork{
+		T: t,
+	}
+}
+
+func (net *testNetwork) Start() {
+	net.createAdvertiser()
```
Comment (iulianpascalau) [inline integrationTests/testNetwork.go:32]: :+1:

**Review by iulianpascalau — APPROVED** [review]

File: integrationTests/testNetwork.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	net.createNodes()
+	net.indexProposers()
+	net.startNodes()
+
+	net.Round = 0
+	net.Nonce = 0
+}
+
+func (net *TestNetwork) Step() {
```
Comment (sasurobert) [inline integrationTests/testNetwork.go:41]: add comment for exported public functions

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: integrationTests/testNetwork.go (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/testNetwork.go:41]: Right, missed that. :D

Comment (camilbancioiu) [reply integrationTests/testNetwork.go]: Added.

**Comment by camilbancioiu** [comment]
Added multiple improvements (see methods of `TestNetwork`).
Added comments.
Rewrote `TestTransaction_TransactionSCScenarios()` to use `TestNetwork` as demonstration.

**Review by SebastianMarian — APPROVED** [review]

File: integrationTests/testNetwork.go [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+
+func (net *TestNetwork) firstNodeInShardOfAddress(address Address) *TestProcessorNode {
+	shardID := net.ShardOfAddress(address)
+	firstNodeInShard := net.NodesSharded[shardID][0]
+	require.NotNil(net.T, firstNodeInShard)
+	return firstNodeInShard
+}
+
+func (net *TestNetwork) handleOrBypassError(err error) {
+	if net.BypassErrorsOnce == true {
```
Comment (SebastianMarian) [inline integrationTests/testNetwork.go:401]: if net.BypassErrorsOnce {

**Review by SebastianMarian — DISMISSED** [review]

File: integrationTests/testNetwork.go [hunk 5]
```diff
@@ -1,107 +1,407 @@
...
+		net.Proposers[i] = i * net.NodesPerShard
 	}
 	net.Proposers[net.NumShards] = net.NumShards * net.NodesPerShard
 }
 
+func (net *TestNetwork) mapNodesByShard() {
+	net.NodesSharded = make(NodesByShardMap)
+	for _, node := range net.Nodes {
+		shardID := node.ShardCoordinator.SelfId()
+		_, exists := net.NodesSharded[shardID]
```
Comment (iulianpascalau) [inline integrationTests/testNetwork.go:351]: L351-L354 can be safely removed

File: integrationTests/testNetwork.go (hunk 4 above)
Comment (camilbancioiu) [reply integrationTests/testNetwork.go:401]: Remnant of a previous condition, removed.

File: integrationTests/testNetwork.go [hunk 6]
```diff
@@ -1,107 +1,407 @@
...
 	"testing"
 
+	"github.com/ElrondNetwork/elrond-go/data/state"
+	"github.com/ElrondNetwork/elrond-go/data/transaction"
 	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/process/factory"
 	"github.com/stretchr/testify/require"
 )
 
+type ShardIdentifier = uint32
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply integrationTests/testNetwork.go:15]: Added.

File: integrationTests/testNetwork.go [hunk 7]
```diff
@@ -1,107 +1,407 @@
...
+	for i := 0; i < count; i++ {
+		shardID := ShardIdentifier(i % net.NumShards)
+		node := net.firstNodeInShard(shardID)
+		net.Wallets[i] = CreateTestWalletAccount(node.ShardCoordinator, shardID)
+	}
+}
+
+// MintWallets adds the specified value to the test wallets.
+func (net *TestNetwork) MintWallets(value *big.Int) {
+	MintAllPlayers(net.Nodes, net.Wallets, value)
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply integrationTests/testNetwork.go:145]: Added TODOs.

File: integrationTests/testNetwork.go (hunk 5 above)
Comment (camilbancioiu) [reply integrationTests/testNetwork.go:351]: Removed.

File: integrationTests/testNetwork.go [hunk 8]
```diff
@@ -1,107 +1,407 @@
...
 	"github.com/stretchr/testify/require"
 )
 
+type ShardIdentifier = uint32
+type Address = []byte
+type NodeSlice = []*TestProcessorNode
+type NodesByShardMap = map[ShardIdentifier]NodeSlice
+type GasScheduleMap = map[string]map[string]uint64
+
+// TODO combine TestNetwork with the preexisting TestContext and OneNodeNetwork
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply integrationTests/testNetwork.go:33]: Fixed.

**Review by iulianpascalau — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**